### PR TITLE
Dockerbuilder: use the arch info from base image

### PR DESCRIPTION
**Pull request by arm64b**
Currently we hardcode the architecture to the `runtime.GOARCH` when building
a docker image, this will result in a confusing info if the arch in the base image is 
different from the one on the host.

This PR takes use of the arch data from the base image during the build process, thus
we can get consistent arch info between the base image and the finally built image.

This PR is trying to resolve one of the defects mentioned in issue #36552.

Signed-off-by: Dennis Chen <dennis.chen@arm.com>

**- What I did**
Make the final generated image has consistent arch info with the base image.
**- How I did it**
Take use of the arch data from the base image.
**- How to verify it**
Using below `Dockerfile` as test:
```
FROM arm32v6/alpine:3.7
CMD ["sh", "-c" , "uname -a"]
```
Then `docker build --no-cache -t test .`
With this PR, we get below output with `docker inspect` on AArch64:
```
...
"Architecture": "arm64",
"Os": "linux",
...
```
After this PR applied, `docker inspect` shows:
```
...
"Architecture": "arm",
"Os": "linux",
...
```
`docker inspect arm32v6/alpine:3.7` shows:
```
...
"Architecture": "arm",
"Os": "linux",
...
```

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Review by vdemeester — APPROVED** [review]
LGTM 🐮

**Review by anusha-ragunathan — APPROVED** [review]
LGTM

**Comment by tonistiigi** [comment]
This can produce wrong results in some cases. For example, if base image is just for some non-platform-specific resources but built on a specific platform (let's say amd64). Now if this image is used in arm with `RUN` commands the resulting image is only valid for arm and not amd64. But with this change the os would be amd64.

**Comment by arm64b** [comment]
Hello @tonistiigi :
>  if base image is just for some non-platform-specific resources but built on a specific platform (let's say amd64)

So this will get a monolithic arch image for amd64-only in this case

>  Now if this image is used in arm with RUN commands the resulting image is only valid for arm and not amd64

That seems you're using the base image for amd64 in arm/arm64 system, can we support this somehow kind of `cross` image build now? I suspect you will encounter problem when pulling the amd64 base image from arm/arm64 system.  Or maybe I don't quite understand well for your comments here, so it will be better if you can give a simple specific example plus what your expectation is 😵  Thanks for the comments!

**Comment by tonistiigi** [comment]
Contrived example for arm.

```
from golang AS build
run go build -o /gocurl ...

from linuxkit/ca-certificates:v0.3-amd64
copy --from=build /gocurl /bin/
```

**Comment by tonistiigi** [comment]
> even without this we still have a equivalent multi-arch one 

This was just an example to make it clear it is `amd64` image. Most of the images are `amd64` and a user has no way to know if it is multi-arch or for a specific arch. And in the case of that image they don't really care because it works on both cases. I don't remember exactly when this happened for linuxkit repos but for official images, it was not long ago when they were not multi-arch.

**Comment by arm64b** [comment]
> Most of the images are amd64 and a user has no way to know if it is multi-arch or for a specific arch

It's the Dockerfile owner's responsibility to select the right base image. Actually we have several ways available can tell you if an image is multi-arch or which arch it is.

> And in the case of that image they don't really care because it works on both cases

Take the example you provide, I don't think the arm-only binary can work well within the container which base img is come from an amd64 arch. You're using `linuxkit/ca-certificates:v0.3-amd64` as the base image and copy arm only binaries into it, what's the purpose to do that, using amd64-certificates to verify the arm binary? I can't understand here.

> update: Another way of looking at it is that a regular distribution manifest (not manifest list) can still be pulled on any arch and it depend on the image itself if it will work. 

That's correct.

> I think this is still the case and if somebody uses this behavior, even without knowing, this would be a breaking change for them.

Seems a wrong causation. If it's breaking, the only reason is the author of the Dockerfile using a wrong arch base image and want to do some tasks that should belong to another different arch, while not this change.

**Comment by kolyshkin** [comment]
> This can produce wrong results in some cases

Yes, but the current code can also produce wrong results, and with this patch the behavior seems to be a safer/saner default (than using the arch dockerd has been built on).

**Comment by kolyshkin** [comment]
In other words, LGTM

**Comment by arm64b** [comment]
Hello @tonistiigi , do you have further comments/suggestions, I think we need to move forward anyway.

**Comment by tonistiigi** [comment]
> do you have further comments/suggestions, I think we need to move forward anyway.

No, that's was the only point I wanted to make. Historically we have had a different behavior where users can rely on using a distribution manifest from other platforms if it doesn't cause runtime errors. If other maintainers want to go ahead with this it is fine by me. Please make sure this gets highlighted in the changelog.

> Yes, but the current code can also produce wrong results, and with this patch the behavior seems to be a safer/saner default (than using the arch dockerd has been built on).

I agree that new behavior is a better default. Unfortunately we have used an incompatible default for 5 years already.

**Comment by thaJeztah** [comment]
ping @tianon @yosifkit @jhowardmsft PTAL

**Review by yosifkit — APPROVED** [review]
Seems sane enough for now. :+1: 

Related: doesn't `OS` need this same logic? (and `variant` and `OSversion` in the future)

**Comment by arm64b** [comment]
> Related: doesn't OS need this same logic? (and variant and OSversion in the future)

I think so! That's also I have another PR #36881 submitted for the preparation of multi-platform support. IOW, it does in the future.

**Comment by arm64b** [comment]
Hello every maintainers, IMO we can merge this PR now, but I don't know if I need to update the changelog as @tonistiigi mentioned...  @thaJeztah @vdemeester, any suggestion about this?

**Review by thaJeztah — APPROVED** [review]
LGTM

don't think I can come up with any useful case where the old behavior would be useful, so let's do this!

**Comment by arm64b** [comment]
Thanks @thaJeztah to push the progress! 👍 For me, technical justification is more important than the PR itself, but anyway we need to move forward IMO😋

**Pull request merged**
